usbSerialForAndroid/src/main/java/com/hoho/android/usbserial/util/SerialInputOutputManager.java

```diff
@@ -155,81 +157,141 @@ public void writeAsync(byte[] data) {
     }
 
     /**
-     * start SerialInputOutputManager in separate thread
+     * start SerialInputOutputManager in separate threads
      */
     public void start() {
-        if(mState != State.STOPPED)
+        if(mState.compareAndSet(State.STOPPED, State.STARTING)) {
+            mStartuplatch = new CountDownLatch(2);
+            mShutdownlatch = new CountDownLatch(2);
+            new Thread(this::runRead, this.getClass().getSimpleName() + "_read").start();
+            new Thread(this::runWrite, this.getClass().getSimpleName() + "_write").start();
+            try {
+                mStartuplatch.await();
+                mState.set(State.RUNNING);
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            }
+        } else {
             throw new IllegalStateException("already started");
-        new Thread(this, this.getClass().getSimpleName()).start();
+        }
     }
 
     /**
-     * stop SerialInputOutputManager thread
+     * stop SerialInputOutputManager threads
      *
      * when using readTimeout == 0 (default), additionally use usbSerialPort.close() to
      * interrupt blocking read
      */
-    public synchronized void stop() {
-        if (getState() == State.RUNNING) {
+    public void stop() {
+        if(mState.compareAndSet(State.RUNNING, State.STOPPING)) {
             Log.i(TAG, "Stop requested");
-            mState = State.STOPPING;
         }
     }
 
-    public synchronized State getState() {
-        return mState;
+    public State getState() {
+        return mState.get();
+    }
+
+    /**
+     * @return true if the thread is still running
+     */
+    private boolean isStillRunning() {
+        State state = mState.get();
+        return ((state == State.RUNNING) || (state == State.STARTING))
+            && (mShutdownlatch.getCount() == 2)
+            && !Thread.currentThread().isInterrupted();
```

Comment:
interrupted wasn't handled so far. Is it really needed?
To my understanding this is never done by the runtime, but here I assume that users unlikely do this on these non-public threads.
I am not familiar with interrupted usage in Java. On Linux an E_INTR is typically handled as retry, not as abort.

Reply:
Yeah, handling Thread.interrupted is good practice in Java. Even if the runtime or users aren't interrupting the thread now, it's a standard mechanism to signal a thread to stop gracefully.